### PR TITLE
Add Exact Match option to Suggest

### PR DESCRIPTION
**Pull request by ricardocerq**
Solves #11579.

Added option to term suggest that returns the input term if it exists. Turned off by default so it doesn't break normal behaviour.

The example query given in #11579 would now be:

```
"Suggest": {
      "term": {
        "field": "word",
        "suggest_mode": "popular",
        "size": 2,
        "prefix_len": 1,
        "analyzer": "default",
        "exact_matching": true
      },
      "text": "Software"
    }
```

And the result:

```
"Suggest": [
      {
         "text": "software",
         "offset": 0,
         "length": 8,
         "options": [
            {
               "text": "software",
               "score": 1,
               "freq": 1
            }
         ]
      }
   ]
```

The output now clearly indicates if the term is present. If it is, its score is always 1, representing an exact match, which differentiates this result from other suggestions.

### CODE REVIEW

**Comment by clintongormley** [comment]
@areek could you have a look at this please.
@ricardocerq you'll also need a test in the PR before it will be merged.

File: core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggester.java
Comment (areek) [inline core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggester.java]: I think we can remove `addExactMatch` method altogether and just use the `TermsEnum#seekExact` to see if the term is present in the term dictionary:
```java
final TermsEnum termsEnum = MultiFields.getTerms(indexReader, token.term.field()).iterator();
if (termsEnum.seekExact(token.term.bytes())) {
    Text word = new StringText(token.term.text());
    resultEntry.addOption(new TermSuggestion.Entry.Option(word, termsEnum.docFreq(), 1f));
}
```

File: core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggester.java
Comment (areek) [inline core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggester.java]: code formatting: ```if (suggestion.getDirectSpellCheckerSettings().exactMatch()) { ``` instead of ```if(suggestion.getDirectSpellCheckerSettings().exactMatch()){```.

File: core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggester.java
Comment (areek) [inline core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggester.java]: See the comment about using `TermsEnum#seekExact` instead of manually iterating through the term dictionary.

File: docs/reference/search/suggesters/term-suggest.asciidoc
Comment (areek) [inline docs/reference/search/suggesters/term-suggest.asciidoc]: we could add `Defaults to False` to indicate the default for this setting

**Comment by areek** [comment]
Thanks @ricardocerq for the PR :). I left some comments.

**Comment by ricardocerq** [comment]
Thanks for the attention and the feedback! I'm making the necessary changes. :)

**Comment by clintongormley** [comment]
@areek could you review this again please?

**Comment by dakrone** [comment]
pinging @areek again for review :)

File: core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggester.java
Comment (areek) [inline core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggester.java]: indentation is off here

**Comment by areek** [comment]
@ricardocerq sorry for the delay, this LGTM, would you mind updating the PR with master?

**Comment by dakrone** [comment]
@ricardocerq are you able to update this by rebasing or merging master in so it can be merged?

**Comment by ricardocerq** [comment]
Sorry about the delay :(. I was unable to retest the changes. If more changes are necessary, please let me know.

**Comment by worldsayshi** [comment]
@areek @clintongormley Can this be merged?

File: core/src/main/java/org/elasticsearch/search/suggest/SuggestUtils.java
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/search/suggest/SuggestUtils.java]: this entire class seems unused?

File: core/src/main/java/org/elasticsearch/search/suggest/DirectSpellcheckerSettings.java
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/search/suggest/DirectSpellcheckerSettings.java]: can we have javadocs on this?

File: modules/lang-mustache/src/test/java/org/elasticsearch/messy/tests/SuggestSearchTests.java
Comment (s1monw) [inline modules/lang-mustache/src/test/java/org/elasticsearch/messy/tests/SuggestSearchTests.java]: this is a full copy of our tests that has been moved back to core into ` org.elasticsearch.search.suggest.SuggestSearchIT` as far as I can tell only `testExactMatch` should be added?

**Review by s1monw — CHANGES_REQUESTED** [review]
the merge wasn't really successful afaik. There are about 1400 lines of unnecessary code in this PR. Otherwise it looks ok but I can't merge it like this. I left comments inline

**Comment by clintongormley** [comment]
@ricardocerq are you still interested in completing this PR?

**Comment by clintongormley** [comment]
Closing in favour of #22902

**Pull request closed without merging**
